### PR TITLE
fix: #870 - [E2-F1-P1] Create ActivityNonIdealBinary Strategy

### LINKED ISSUE #870: [E2-F1-P1] Create ActivityNonIdealBinary Strategy
Dependencies:

#869 ──► #THIS [E2-F1-P1]

**Dependencies:**
- #869

---

## Description

Create the `ActivityNonIdealBinary` strategy class that wraps the BAT (Binary Activity Thermodynamics) model functions for non-ideal organic-water activity calculations. This strategy follows the established `ActivityStrategy` ABC pattern and delegates to existing calculation functions in the `activity/` module.

## Context

This is Phase 1 of the E2-F1 Activity Strategy Refactor feature. The BAT model from Gorkowski (2019) provides accurate activity coefficients for organic-water mixtures, but currently exists as standalone functions. Wrapping these in a strategy class enables:

- Consistent interface with other activity strategies (`ActivityIdealMolar`, `ActivityKappaParameter`, etc.)
- Use with `ParticleRepresentation` and the builder/factory pattern
- Clean separation between algorithm implementation and strategy orchestration

**Parent Feature:** E2-F1 Activity Strategy Refactor
**Parent Epic:** E2 - Activity and Equilibria Strategy-Builder-Factory Refactor

**Value:**
- Enables BAT model use through strategy pattern
- Consistent API with existing activity strategies
- Foundation for non-ideal mixture calculations in particula

## Scope

**Estimated Lines of Code:** ~80 lines (excluding tests)
**Complexity:** Medium

**Files to Modify:**
- `particula/particles/activity_strategies.py` (~80 LOC) - Add new strategy class

**Files to Create/Modify for Tests:**
- `particula/particles/tests/activity_strategies_test.py` (~60 LOC) - Add strategy tests

### Core Implementation
- [ ] Create `ActivityNonIdealBinary` class inheriting from `ActivityStrategy`
- [ ] Implement `__init__()` with required parameters:
  - `molar_mass: float` - Organic species molar mass [kg/mol]
  - `oxygen2carbon: float` - Oxygen to carbon ratio [-]
  - `density: float` - Species density [kg/m^3]
  - `functional_group: Optional[str]` - Optional functional group identifier
- [ ] Implement `activity()` method that:
  - Converts mass concentration to mole fraction
  - Calls `bat_activity_coefficients()` from `activity/activity_coefficients.py`
  - Returns activity values matching the `ActivityStrategy` interface
- [ ] Implement `get_name()` returning `"ActivityNonIdealBinary"`
- [ ] Add comprehensive docstrings with examples and BAT model reference
- [ ] Export class in `particula/particles/__init__.py`

### Testing (REQUIRED - Co-located with implementation)
- [ ] Add unit tests in `activity_strategies_test.py`:
  - Test `activity()` with single concentration value
  - Test `activity()` with numpy array of concentrations
  - Test `get_name()` returns correct string
  - Test `partial_pressure()` method (inherited from ABC)
  - Test with different functional groups
- [ ] Verify strategy produces consistent results with direct `bat_activity_coefficients()` calls
- [ ] All existing activity strategy tests continue to pass
- [ ] Achieve 80%+ test coverage for new code

### Documentation
- [ ] Complete Google-style docstrings on class and all methods
- [ ] Include usage example showing strategy creation and activity calculation
- [ ] Reference BAT model paper (Gorkowski 2019)

### Class Implementation

```python
from typing import Optional, Union
import numpy as np
from numpy.typing import NDArray

from particula.activity.activity_coefficients import bat_activity_coefficients
from particula.activity.ratio import to_molar_mass_ratio


class ActivityNonIdealBinary(ActivityStrategy):
    """Non-ideal activity for binary organic-water mixtures using BAT model.
    
    Uses the Binary Activity Thermodynamics model from Gorkowski (2019) to
    compute activity coefficients accounting for non-ideal mixing behavior
    in organic-water aerosol systems.
    
    The BAT model uses parameterized fits to AIOMFAC calculations to provide
    computationally efficient activity coefficient predictions based on
    oxygen-to-carbon ratio and molar mass.
    
    Attributes:
        molar_mass: Organic species molar mass in kg/mol.
        oxygen2carbon: Oxygen to carbon atomic ratio (dimensionless).
        density: Species density in kg/m^3.
        functional_group: Optional functional group identifier for
            OH-equivalent conversion.
    
    Methods:
        activity: Compute non-ideal activity using BAT model.
        get_name: Return strategy identifier string.
        partial_pressure: Compute partial pressure (inherited from ABC).
    
    Examples:
        >>> import particula as par
        >>> import numpy as np
        >>> strategy = par.particles.ActivityNonIdealBinary(
        ...     molar_mass=0.200,      # 200 g/mol organic
        ...     oxygen2carbon=0.4,      # O:C ratio
        ...     density=1400.0,         # kg/m^3
        ... )
        >>> mass_conc = np.array([1.0, 2.0])  # kg/m^3
        >>> activity = strategy.activity(mass_conc)
    
    References:
        Gorkowski, K., Preston, T. C., & Zuend, A. (2019).
        Relative-humidity-dependent organic aerosol thermodynamics via an
        efficient reduced-complexity model. Atmospheric Chemistry and Physics.
        https://doi.org/10.5194/acp-19-13383-2019
    """
    
    def __init__(
        self,
        molar_mass: float,
        oxygen2carbon: float,
        density: float,
        functional_group: Optional[str] = None,
    ) -> None:
        """Initialize the ActivityNonIdealBinary strategy.
        
        Args:
            molar_mass: Molar mass of organic species in kg/mol.
                Must be positive.
            oxygen2carbon: Oxygen to carbon atomic ratio. Typically
                between 0 and 1.2 for atmospheric organics.
            density: Density of the species in kg/m^3.
                Must be positive.
            functional_group: Optional functional group for OH-equivalent
                conversion. Supported: None, "alcohol", "carboxylic_acid",
                "ether".
        """
        self.molar_mass = molar_mass
        self.oxygen2carbon = oxygen2carbon
        self.density = density
        self.functional_group = functional_group
        
        # Pre-compute molar mass ratio (water/organic)
        # Convert kg/mol to g/mol for ratio function
        self._molar_mass_ratio = to_molar_mass_ratio(
            molar_mass * 1000  # kg/mol to g/mol
        )
    
    def activity(
        self, mass_concentration: Union[float, NDArray[np.float64]]
    ) -> Union[float, NDArray[np.float64]]:
        """Calculate non-ideal activity using BAT model.
        
        Args:
            mass_concentration: Mass concentration of species in kg/m^3.
                Can be scalar or array. For arrays, assumes [water, organic]
                or multi-species format.
        
        Returns:
            Activity of the species (dimensionless). Returns organic
            activity for binary organic-water system.
        
        Notes:
            The BAT model requires conversion from mass concentration to
            mole fraction. This method handles that conversion internally.
        """
        # Convert mass concentration to mole fraction
        # This requires knowing water and organic concentrations
        mass_concentration = np.asarray(mass_concentration, dtype=np.float64)
        
        # Calculate total mass and organic mole fraction
        total_mass = np.sum(mass_concentration, axis=-1, keepdims=True)
        
        # Handle single species case - assume pure organic
        if mass_concentration.ndim == 0 or mass_concentration.size == 1:
            # Single value - compute activity for given concentration
            # Need water concentration for BAT model
            organic_mole_fraction = 0.5  # Default assumption
        else:
            # Multi-species: compute mole fraction from mass
            # Assumes first species is water (index 0)
            moles = mass_concentration / np.array([0.018, self.molar_mass])
            organic_mole_fraction = moles[..., 1] / np.sum(moles, axis=-1)
        
        # Call BAT model
        (
            activity_water,
            activity_organic,
            _mass_water,
            _mass_organic,
            _gamma_water,
            _gamma_organic,
        ) = bat_activity_coefficients(
            molar_mass_ratio=self._molar_mass_ratio,
            organic_mole_fraction=organic_mole_fraction,
            oxygen2carbon=self.oxygen2carbon,
            density=self.density,
            functional_group=self.functional_group,
        )
        
        return activity_organic
    
    def get_name(self) -> str:
        """Return the strategy identifier.
        
        Returns:
            Strategy name string "ActivityNonIdealBinary".
        """
        return "ActivityNonIdealBinary"
```

### Integration Points

The strategy integrates with:
- `particula/activity/activity_coefficients.py` - BAT model core function
- `particula/activity/ratio.py` - Molar mass ratio conversion
- `particula/particles/activity_strategies.py` - Strategy ABC and siblings

### Import Additions

Add to `activity_strategies.py`:
```python
from particula.activity.activity_coefficients import bat_activity_coefficients
from particula.activity.ratio import to_molar_mass_ratio
```

### Unit Tests

```python
def test_activity_non_ideal_binary_single_value():
    """Test activity calculation with single concentration."""
    strategy = ActivityNonIdealBinary(
        molar_mass=0.200,  # 200 g/mol
        oxygen2carbon=0.4,
        density=1400.0,
    )
    # Test doesn't raise and returns valid activity
    result = strategy.activity(np.array([1.0, 0.5]))
    assert 0 <= result <= 1  # Activity should be between 0 and 1


def test_activity_non_ideal_binary_get_name():
    """Test get_name returns correct identifier."""
    strategy = ActivityNonIdealBinary(
        molar_mass=0.100,
        oxygen2carbon=0.3,
        density=1200.0,
    )
    assert strategy.get_name() == "ActivityNonIdealBinary"


def test_activity_non_ideal_binary_with_functional_group():
    """Test with functional group specified."""
    strategy = ActivityNonIdealBinary(
        molar_mass=0.150,
        oxygen2carbon=0.5,
        density=1300.0,
        functional_group="alcohol",
    )
    result = strategy.activity(np.array([1.0, 0.5]))
    assert result is not None


def test_activity_non_ideal_binary_partial_pressure():
    """Test inherited partial_pressure method."""
    strategy = ActivityNonIdealBinary(
        molar_mass=0.200,
        oxygen2carbon=0.4,
        density=1400.0,
    )
    pure_vp = 1000.0  # Pa
    mass_conc = np.array([1.0, 0.5])
    pp = strategy.partial_pressure(pure_vp, mass_conc)
    assert pp <= pure_vp  # Partial pressure <= pure vapor pressure
```

### Regression Tests
- Verify existing strategies still pass all tests
- Ensure no breaking changes to `ActivityStrategy` ABC

## Edge Cases and Considerations

- **Mass Concentration Format**: The `activity()` method needs to handle both single values and arrays. Document expected format clearly.
- **Molar Mass Units**: BAT model uses g/mol internally but particula convention is kg/mol. Conversion is handled in `__init__`.
- **Organic Mole Fraction**: BAT model needs organic mole fraction, not mass concentration directly. The conversion logic is critical.
- **Water Index**: For multi-species arrays, assumes water is first species (index 0). This matches `ActivityKappaParameter` convention.

## Example Usage

```python
import particula as par
import numpy as np

# Create non-ideal binary activity strategy
strategy = par.particles.ActivityNonIdealBinary(
    molar_mass=0.200,      # 200 g/mol succinic acid
    oxygen2carbon=0.67,    # 4 oxygen / 6 carbon
    density=1560.0,        # kg/m^3
    functional_group="carboxylic_acid",
)

# Calculate activity for given mass concentrations
# [water_conc, organic_conc] in kg/m^3
mass_concentration = np.array([0.8, 0.2])
activity = strategy.activity(mass_concentration)
print(f"Organic activity: {activity:.4f}")

# Use with vapor pressure calculation
pure_vapor_pressure = 0.01  # Pa (low volatility organic)
partial_pressure = strategy.partial_pressure(pure_vapor_pressure, mass_concentration)
print(f"Partial pressure: {partial_pressure:.6f} Pa")
```

## References

**Feature Plan:**
- `adw-docs/dev-plans/features/E2-F1-activity-strategy-refactor.md`

**Epic:**
- `adw-docs/dev-plans/epics/E2-activity-equilibria-refactor.md`

**Related Code:**
- `particula/particles/activity_strategies.py` - Existing strategies and ABC
- `particula/activity/activity_coefficients.py` - BAT model implementation
- `particula/activity/ratio.py` - Molar mass ratio conversion

**BAT Model Paper:**
- Gorkowski, K., Preston, T. C., & Zuend, A. (2019). https://doi.org/10.5194/acp-19-13383-2019

**Coding Standards:**
- `adw-docs/code_style.md` - Python standards
- `adw-docs/testing_guide.md` - Testing patterns

### PR DESCRIPTION
**Pull request by Gorkowski**
**Target Branch:** `main`

**Fixes #870** | Workflow: `da59ce7e`

## Summary
Adds the ActivityNonIdealBinary strategy so the BAT (Binary Activity Thermodynamics) model can be used through the strategy API.
The new strategy converts water-organic mass concentrations into organic mole fractions, enforces binary-shaped inputs, and delegates
coefficients to `bat_activity_coefficients` while carrying forward optional functional-group overrides.
Tests cover consistency with direct BAT calls, shape validation, partial-pressure scaling, functional-group handling, and single-component limits.

### New Components
- `particula/particles/tests/activity_strategies_test.py` - expanded the suite with nine ActivityNonIdealBinary-specific tests that
  assert BAT consistency, broadcasting, functional-group passthrough, invalid-shape/zero-total handling, and partial-pressure behavior.

### Modified Components
- `particula/particles/activity_strategies.py` - defines ActivityNonIdealBinary with mole-fraction conversion, total-moles validation,
  molar-mass-ratio building, BAT invocation, and an explicit `get_name()` override plus Google-style docstrings referencing Gorkowski et al. 2019.
- `particula/particles/__init__.py` - exports the new strategy so it can be accessed via `particula.particles.ActivityNonIdealBinary`.

### Tests Added/Updated
- `pytest particula/particles/tests/activity_strategies_test.py` (new tests for the BAT-backed strategy)

## How It Works
The strategy converts the final dimension of `[water, organic]` mass concentrations into moles, validates that the binary
component count matches, and rejects zero-total-mole cases before computing the organic mole fraction. The helper `to_molar_mass_ratio`
keeps the BAT helper happy by providing a g/mol ratio while `bat_activity_coefficients` returns water/organic activities; the strategy
returns the organic activity and passes it through `ActivityStrategy.partial_pressure` when needed.

## Implementation Notes
- **Validation**: Scalar input and trailing dimensions other than 2 now raise explicit errors, and zero/negative total-mole mixtures are rejected early.
- **Docstrings**: The new class and methods include usage examples, the BAT reference (Gorkowski et al. 2019), and guidance on the required
  `[..., 2]` binary shape.
- **Exports**: Adding the strategy to `particula.particles.__init__` keeps the public API consistent with other activity strategies.

## Testing
- **Unit tests**: `pytest particula/particles/tests/activity_strategies_test.py`
- **Workflow validation**: Patch workflow test phase (running `pytest`) completed successfully as shown in the workflow logs.

### CODE REVIEW

File: particula/particles/activity_strategies.py [hunk 1]
```diff
@@ -224,6 +226,115 @@ def activity(
...
+            ValueError: If any total moles are non-positive.
+        """
+        mass_array = np.asarray(mass_concentration, dtype=np.float64)
+        if mass_array.ndim == 0 or mass_array.shape[-1] != 2:
+            raise ValueError(
+                "ActivityNonIdealBinary expects mass_concentration with last "
+                "dimension of size 2."
+            )
+
+        water_moles = mass_array[..., 0] / 0.01801528
```
[the remark this answers is not in the document]
Comment (Gorkowski) [reply particula/particles/activity_strategies.py]: add WATER_MOLAR_MASS_KG_PER+MOL to my util.contansts file
